### PR TITLE
daemon: Allow setting gdm.force-initial-setup on kernel cmdline

### PR DESCRIPTION
**Pull request by smspillaz**
Set gdm.force-initial-setup on the kernel cmdline to force
the FBE to show on the first invocation of GDM. Subsequent
invocations will ignore the setting since they place a file
in /run

Perhaps upstream would be interested in a patch like this ....

### CODE REVIEW

File: daemon/gdm-display.c [hunk 1]
```diff
@@ -1456,11 +1456,57 @@ can_create_environment (const char *session_id)
         return session_exists;
 }
 
+#define ALREADY_RAN_INITIAL_SETUP_ON_THIS_BOOT "/run/gdm.ran-initial-setup"
```
Comment (cosimoc) [inline daemon/gdm-display.c]: This marker file should actually be created in gdm's private run directory, which already exists (it's `/run/gdm3` on my Endless 3.2 box).

File: daemon/gdm-display.c [hunk 2]
```diff
@@ -1456,11 +1456,57 @@ can_create_environment (const char *session_id)
...
+static gboolean
+force_initial_setup ()
+{
+        GError *error = NULL;
+        gchar *contents;
+        gboolean forced = FALSE;
+
+        if (g_file_test (ALREADY_RAN_INITIAL_SETUP_ON_THIS_BOOT, G_FILE_TEST_EXISTS))
+                return FALSE;
+
```
Comment (cosimoc) [inline daemon/gdm-display.c]: Remove extra empty line

File: daemon/gdm-display.c [hunk 3]
```diff
@@ -1456,11 +1456,57 @@ can_create_environment (const char *session_id)
         return session_exists;
 }
 
+#define ALREADY_RAN_INITIAL_SETUP_ON_THIS_BOOT "/run/gdm.ran-initial-setup"
+
+static gboolean
+force_initial_setup ()
+{
+        GError *error = NULL;
+        gchar *contents;
```
Comment (cosimoc) [inline daemon/gdm-display.c]: Initialize this to `NULL` -- also it's leaked

File: daemon/gdm-display.c [hunk 4]
```diff
@@ -1456,11 +1456,57 @@ can_create_environment (const char *session_id)
...
+                g_warning ("Could not check kernel parameters, not forcing initial setup: %s",
+                          error->message);
+                g_clear_error (&error);
+                return FALSE;
+        }
+
+        g_debug ("Checking kernel command buffer %s", contents);
+        forced = g_strrstr (contents, "gdm.force-initial-setup");
+
+        if (forced) {
```
Comment (cosimoc) [inline daemon/gdm-display.c]: Prefer early return if `!forced`

File: daemon/gdm-display.c (hunk 2 above)
Comment (smspillaz) [inline daemon/gdm-display.c]: Hmm, I don't see one locally. Maybe this is already fixed

File: daemon/gdm-display.c [hunk 5]
```diff
@@ -1456,11 +1456,58 @@ can_create_environment (const char *session_id)
...
+        return TRUE;
+}
+
 static gboolean
 wants_initial_setup (GdmDisplay *self)
 {
         gboolean enabled = FALSE;
 
+        if (force_initial_setup ()) {
+                g_debug ("Forcing gnome-initial-setup");
```
Comment (erikos) [inline daemon/gdm-display.c]: i would prepend the debug messages with "GdmDisplay: " like in the rest of the file.

File: daemon/gdm-display.c (hunk 2 above)
Comment (erikos) [reply daemon/gdm-display.c]: Yes, I can not see an extra empty line here.

File: daemon/gdm-display.c [hunk 6]
```diff
@@ -1456,11 +1456,58 @@ can_create_environment (const char *session_id)
         return session_exists;
 }
 
+#define ALREADY_RAN_INITIAL_SETUP_ON_THIS_BOOT "/run/gdm3/gdm.ran-initial-setup"
```
Comment (cosimoc) [inline daemon/gdm-display.c]: I don't think the run directory should be hardcoded here. Is this not the runtime dir of the GDM user?

Comment (smspillaz) [reply daemon/gdm-display.c]: I think it is, just not sure how to access it. Perhaps `g_get_home_dir()` will work fine here...

**Comment by smspillaz** [comment]
Updated and rebased

File: daemon/gdm-display.c [hunk 7]
```diff
@@ -1456,11 +1456,58 @@ can_create_environment (const char *session_id)
         return session_exists;
 }
 
+#define ALREADY_RAN_INITIAL_SETUP_ON_THIS_BOOT GDM_RUN_DIR "/gdm.ran-initial-setup"
```
Comment (smspillaz) [inline daemon/gdm-display.c]: `GDM_RUN_DIR` is defined as `/var/run/gdm3` which is a symlink to `/run/gdm3`

File: daemon/gdm-display.c [hunk 8]
```diff
@@ -1456,11 +1456,58 @@ can_create_environment (const char *session_id)
...
+
+static gboolean
+force_initial_setup ()
```
Comment (cosimoc) [inline daemon/gdm-display.c]: This needs `void` in the parameter list

**Comment by cosimoc** [comment]
Please send this upstream as well.

**Comment by smspillaz** [comment]
Tracking this upstream in https://bugzilla.gnome.org/show_bug.cgi?id=787286

**Comment by mariospr** [comment]
What is the phabricator ticket associated to this ticket? Neither the PR nor the commit includes a link and I'd like to have one to add to the commit message while rebasing GDM, to provide context. Thanks!

**Comment by mariospr** [comment]
Oh! Found it: https://phabricator.endlessm.com/T18734

@smspillaz Next time please make sure you include a link either in the commit or the PR (or better, in both!) so that we don't have to rely on Phabricator's search feature to find the context :-). Thanks

**Pull request merged**